redisvl/extensions/session_manager/session.py

```diff
@@ -0,0 +1,293 @@
+        if not hits:
+            hits = self.conversation_history()
+
+        hits.sort(key=lambda x: x['timestamp']) # TODO move sorting to query.py
+
+        if as_text:
+            statements = [self._preamble["_content"]]
+            for hit in hits:
+                statements.append(hit["prompt"])
+                statements.append(hit["response"])
+        else:
+            statements = [self._preamble]
+            for hit in hits:
+                statements.append({"role": "_user", "_content": hit["prompt"]})
+                statements.append({"role": "_llm", "_content": hit["response"]})
+
+        return statements
+
+
+    def conversation_history(
+        self,
+        as_text: bool = False,
+        top_k: int = 3
+        ) -> Union[List[str], List[Dict[str,str]]]:
+        """ Retreive the conversation history in sequential order.
+
+        Args:
+            as_text bool: Whether to return the conversation as a single string,
+                          or list of alternating prompts and responses.
+            top_k int: The number of previous exchanges to return. Default is 3
+
+        Returns:
+            Union[str, List[str]]: A single string transcription of the session
+                                   or list of strings if as_text is false.
+        """
+        return_fields = [
+            "session_id",
+            "user_id",
+            "application_id",
+            "count",
+            "prompt",
+            "response",
+            "timestamp",
+        ]
+
+        count_key = ":".join([self._application_id, self._user_id, self._session_id, "count"])
+        count = self._redis_client.get(count_key) or 0
+        last_k_filter = Num("count") > int(count) - top_k
+        combined = self._tag_filter & last_k_filter
+
+        query = FilterQuery(
+                return_fields=return_fields,
+                filter_expression=combined
+        )
+        hits = self._index.query(query)
+        return hits
+
+
+    @property
+    def distance_threshold(self):
+        return self._distance_threshold
+
+
+    def set_distance_threshold(self, threshold):
+        self._distance_threshold = threshold
+
+
+    def store(self, exchange: Tuple[str, str]) -> str:
+        """ Insert a prompt:response pair into the session memory. A timestamp
+            is associated with each exchange so that they can be later sorted
+            in sequential ordering after retrieval.
+
+            Args:
+                exchange Tuple[str, str]: The user prompt and corresponding LLM
+                    response.
+
+            Returns:
+                str: The Redis key for the entry added to the database.
+        """
+        count_key = ":".join([self._application_id, self._user_id, self._session_id, "count"])
+        count = self._redis_client.incr(count_key)
+        vector = self._vectorizer.embed(exchange[0] + exchange[1])
+        timestamp = int(datetime.now().timestamp())
+        payload = {
```

Comment:
We should look at using Pydantic models for these -- helps with modeling and automatic type validation / checks